### PR TITLE
Refactoring before adding GitHub action support

### PR DESCRIPTION
**Pull request by kevmoo**
Added a shared library for CI code.
Moved a lot of code out of `lib/src/travis/travis_shell.dart` -> `lib/src/ci_test_script.dart`
Renamed `TravisJob` to `CIJob`
Renamed `_TravisJobEntry` to `CIJobEntry`

### CODE REVIEW

File: mono_repo/lib/src/ci_shared.dart [hunk 1]
```diff
@@ -14,6 +16,60 @@ String createdWith() => Zone.current[skipCreatedWithSentinel] == true
...
+
+  @override
+  bool operator ==(Object other) =>
+      other is CIJobEntry &&
+      _equality.equals(_identityItems, other._identityItems);
+
+  @override
+  int get hashCode => _equality.hash(_identityItems);
+
+  List get _identityItems => [job.os, job.stageName, job.sdk, commands, merge];
```
Comment (jakemac53) [inline mono_repo/lib/src/ci_shared.dart:41]: It seems like we should probably at least have a comment here explaining why we don't use all the fields from the CIJob (or its ==/hashCode)

File: mono_repo/lib/src/ci_shared.dart [hunk 2]
```diff
@@ -14,6 +16,60 @@ String createdWith() => Zone.current[skipCreatedWithSentinel] == true
...
+void validateRootConfig(RootConfig rootConfig) {
+  for (var config in rootConfig) {
+    final sdkConstraint = config.pubspec.environment['sdk'];
+
+    if (sdkConstraint == null) {
+      continue;
+    }
+
+    final disallowedExplicitVersions = config.jobs
+        .map((tj) => tj.explicitSdkVersion)
```
Comment (jakemac53) [inline mono_repo/lib/src/ci_shared.dart:53]: s/tj/job

File: mono_repo/lib/src/ci_shared.dart [hunk 3]
```diff
@@ -14,6 +16,60 @@ String createdWith() => Zone.current[skipCreatedWithSentinel] == true
...
+        .where((v) => v != null)
```
Comment (jakemac53) [inline mono_repo/lib/src/ci_shared.dart:54]: nit: `&& !sdkConstraint.allows(v)` and remove the extra where etc? speed should not be a concern here

File: mono_repo/lib/src/ci_shared.dart [hunk 4]
```diff
@@ -0,0 +1,85 @@
+import 'dart:async';
```
[the remark this answers is not in the document]
Comment (kevmoo) [reply mono_repo/lib/src/ci_shared.dart:1]: ack

File: mono_repo/lib/src/ci_shared.dart (hunk 1 above)
Comment (kevmoo) [reply mono_repo/lib/src/ci_shared.dart]: You made me do a bigger fix here, which is good.

File: mono_repo/lib/src/ci_shared.dart (hunk 2 above)
Comment (kevmoo) [reply mono_repo/lib/src/ci_shared.dart:41]: yup

File: mono_repo/lib/src/ci_shared.dart (hunk 3 above)
Comment (kevmoo) [reply mono_repo/lib/src/ci_shared.dart:42]: 🤷

File: mono_repo/lib/src/commands/travis/travis_yaml.dart [hunk 5]
```diff
@@ -209,72 +210,56 @@ List<Object> _calculateOrderedStages(RootConfig rootConfig) {
...
         job.tasks.map((task) => commandsToKeys[task.command]).toList();
 
-    jobEntries.add(
-        _TravisJobEntry(job, commands, mergeStages.contains(job.stageName)));
+    jobEntries.add(CIJobEntry(job, commands));
   }
 
-  final groupedItems =
-      groupBy<_TravisJobEntry, _TravisJobEntry>(jobEntries, (e) => e);
+  // Group jobs by all of the values that would allow them to merge
```
Comment (kevmoo) [inline mono_repo/lib/src/commands/travis/travis_yaml.dart:226]: @jakemac53 – I was using that crazy equality mess to handle this case. Instead, I just inlined the logic.

**Review by jakemac53 — APPROVED** [review]

**Pull request merged**